### ISSUE TITLE
java.lang.IllegalArgumentException: View not attached to window manager

### ISSUE THREAD

**Issue by donpwilson** [CloudRail/cloudrail-si-android-sdk#26]
I am implementing CloudRail with all calls wrapped within android-priority-jobqueue Jobs. These Jobs are completely independent of all Activities so that the user can carry on with other work in the app while the upload/download is proceeding. I therefore initialize the Services with the application context. I have successfully searched cloud storage for all photos and downloaded a selection using Glide.

However, I am getting errors:
Activity com.cloudrail.si.servicecode.commands.awaitCodeRedirect.AuthenticationActivity has leaked window com.android.internal.policy.impl.PhoneWindow$DecorView{4244f300 V.E..... R......D 0,0-563,96} that was originally added here
                                                                    android.view.WindowLeaked: Activity com.cloudrail.si.servicecode.commands.awaitCodeRedirect.AuthenticationActivity has leaked window com.android.internal.policy.impl.PhoneWindow$DecorView{4244f300 V.E..... R......D 0,0-563,96} that was originally added here
                                                                        at android.view.ViewRootImpl.<init>(ViewRootImpl.java:454)
                                                                        at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:235)
                                                                        at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:73)
                                                                        at android.app.Dialog.show(Dialog.java:282)
                                                                        at android.app.ProgressDialog.show(ProgressDialog.java:116)
                                                                        at android.app.ProgressDialog.show(ProgressDialog.java:99)
                                                                        at com.cloudrail.si.servicecode.commands.awaitCodeRedirect.AuthenticationActivity.onCreate(AuthenticationActivity.java:25)
                                                               
which do not crash the app.
I also get intermittent errors:

java.lang.IllegalArgumentException: View not attached to window manager
                                                                         at android.view.WindowManagerGlobal.findViewLocked(WindowManagerGlobal.java:402)
                                                                         at android.view.WindowManagerGlobal.removeView(WindowManagerGlobal.java:304)
                                                                         at android.view.WindowManagerImpl.removeView(WindowManagerImpl.java:83)
                                                                         at android.app.Dialog.dismissDialog(Dialog.java:325)
                                                                         at android.app.Dialog.dismiss(Dialog.java:307)
                                                                         at com.cloudrail.si.servicecode.commands.awaitCodeRedirect.AuthenticationActivity$1.onPageFinished(AuthenticationActivity.java:60)
                                                                    
which do crash the app. Apologies for the lack of formatting above.

These errors appear to come from the progress dialog (full screen in gray, with a Loading progress bar) which appears for a few seconds at several points.

Ideally, I would not have a progress dialog shown at all because it gives little meaningful information and, behind it, my app indicates progress to the user.

**Comment by PatrickStoklasa** [comment]
The ProgressDialog is started whenever the authentication process is triggered and the WebView is loading the login page. It is closed as soon as it has been loaded completely. The reason for adding it was that on devices with a slow internet connection it only showed a white screen while it was loading. Since it can take a few seconds to load the website in that case, a lot of users pressed the back button because they didn't see any progress.

It appears that it is trying to close the ProgressDialog after the Activity has been shut down already. Google does not provide a nice way to check that before calling dismiss() on a Dialog. Nonetheless, we will try to find a workaround for your issue.

In the meantime, how are you handling authentication? Do you use the explicit way by calling login() first? If not, are you trying to do multiple calls at the same time such that the authentication process might be started multiple times?

**Comment by donpwilson** [comment]
Thank you for your quick response.

I am working from the code in the sample cloud storage app as follows. I initialize once in a singleton class. This takes care of any login required. In my testing which threw up the errors I only had to login to OneDrive once. Thereafter the login page was bypassed. Abbreviated code:

    private final AtomicReference<CloudStorage> onedrive = new AtomicReference<>();

    onedrive.set(new OneDrive(context, Get.GetString(R.string.one_drive_id), Get.GetString(R
                    .string.one_drive_secret)));
       
(Note Get.GetString is a helper method I use in my app)

I then call for a service from wherever I need it in my app:

  public CloudStorage getService(final int drive) {
        switch (drive) {
            case My_Constants.DRIVE_ONE_DRIVE:
                return onedrive.get();
              default:
                return null;
        }
    }

I have not yet implemented the persistent saveAsString() and loadAsString() methods because I want to take it one step at a time and fully understand what it going on.

It would appear from your description that the authentication process is being triggered multiple times - because the ProgressDialog appears multiple times - but I can't see why this should be happening. My internet connection is quite slow and is periodically lost. Does it being recovered trigger authentication again? Also. what is "the explicit way by calling login() first" that you refer to? Is that what new (OneDrive) does?

The error on trying to close the ProgressDialog occurs some time after the dialog has actually closed on the screen so, maybe, it is being called multiple times.

To sum up:
- the ProgressDialog makes sense
- I need this dialog to be closed cleanly
- we need to find a way of avoiding attempting the authentication process multiple times

I'll continue to experiment but await your further comments, thanks

**Comment by PatrickStoklasa** [comment]
We just pushed a new version (2.6.2) which should solve the error you get with the ProgressDialog. Regarding the explicit login i mentioned: Our library provides a methods called login for the cloud storage services with which you can trigger the authentication process manually. In order to solve your problem you could call login whenever a user selects one of the cloud storage providers or just at a fixed point in your application and only when this method was executed correctly do any further calls to the library. This ensures that the login process is triggered first and only ones if done correctly.

**Comment by donpwilson** [comment]
I confirm that the new version clears both errors and my app is now running. I have also implemented explicit login before any call for a service. However, I still get a couple of additional flashes of the dialog (with a white rather than grey background) which I am investigating. Do the different background colours signify different dialogs?

Worse though, while downloading multi Mbyte photos for a few seconds I got a succession of "ServiceCode Error in function checkAuthentication at 3" errors. I am assuming that these were transmission errors because Glide successfully resolved them and loaded the photos. However, each error message was accompanied by a brief flash of a dialog. Not very user friendly. This should all be "handled behind the scenes".

I still find the ProgressDialog rather uninformative. Would it be possible for it to have a transparent background. It would then be possible for my app (and other developers' apps) to provide a fuller / better explanation if they wish.

Thanks

**Comment by PatrickStoklasa** [comment]
The only reasons why you might get multiple flashes of the screen are either that you try to make multiple calls at the same time or that you reinitialize the service every time you do a new call. Without additional source code this will be hard to tell for me. 

The second thing you are describing is most likely because of the first one since we only display something during the authentication process which should not be triggered at this place. We will investigate this further if it persists after fixing the main issue.

Changing the background of the ProgressDialog wouldn't allow you to display anything since we run a new activity that displays a WebView together with the ProgressDialog so you wouldn't be able to add any UI elements there. What would you prefer as a text on the ProgressDialog instead?

**Comment by donpwilson** [comment]
With regard to the multiple flashes of the screen, I'll continue to flesh out my app with the CloudRail implementation and monitor whether it continues to occur and, if so, in what circumstances. I've had to re-code some elements for instance because closure of the WebView triggers a call to onResume() in my activity. I still suspect that it is happening either if the internet connection is lost, or is regained.

Ideally the ProgressDialog should display something like "Connecting to [service] ..." where service is "Google Drive", "OneDrive", or whatever.

**Comment by donpwilson** [comment]
You will be pleased to know that I no longer get multiple flashes of the ProgressDialog screen.

By my analysis there are three possibilities when a (implicit) login is required:
1. the user is already logged in
2. the user is not logged in but the credentials are loaded into the service so the login can happen automatically
3. the user is not logged in and there are no stored credentials so the login dialog needs to be displayed

As fas as I can see no ProgressDialog is ever displayed while 1. is being checked. Good!

The ProgressDialog is displayed for 3, followed by the login view. OK!

The ProgressDialog is displayed for 2, for several seconds in my testing. I don't believe that it should be displayed in this circumstance.  When the login is being done in the background there should not be a need for a WebView page so there should be no need to block the ui, with the current activity being paused, etc. If there really is a need for a a WebView page then it would make more sense is it was displayed briefly when loaded showing the completed entries, replacing the ProgressDialog.

Do you agree, or have I missed something?

**Comment by PatrickStoklasa** [comment]
You are right, in the second case there should not be a ProgressDialog neither a WebView involved! I will check it again and see if i get the same behavior. Does this happen for all of the services or only for a few / one?

**Comment by donpwilson** [comment]
It certainly happens with Dropbox.  I'm not yet using Google Drive. For Box I'm not sure because it keeps on asking me to log in again. I haven't yet isolated this as an issue or not. It might be related to #29.

I'm not sure with OneDrive because I'm not sure that OneDrive actually allows users to logout. My research online suggests that logging out of OneDrive is extremely difficult and can only be achieved by withdrawing permission in OneDrive settings. In my app, if there are no credentials in the service, the ProgressDialog is displayed, but no login dialog, presumably because it then finds that I'm already logged in. This suggests that the test for displaying the ProgressDialog should not just be for whether the credentials are held but also whether the user is already logged in. In the latter case, the credentials used for the login should be stored into the service.

fyi, in my app there is an option to logout of any drive. This calls service.logout() and at the same time it clears the persistent values - which I store in Preferences.

**Comment by donpwilson** [comment]
Godd news! I've found and cleared some bugs / logic errors in my app which has cleared the issue I was experiencing with the ProgressDialog appearing when it shouldn't. I'm left with OneDrive not logging out when requested and it would be great if you could comfirm my analysis above that OneDrive does not allow logout.

To confirm, both Dropbox and Box are working fine. Enhanced wording for the dialog would still be useful.

**Comment by PatrickStoklasa** [comment]
You are right about OneDrive logout. Since they don't provide a method for removing access, we only clear the token that the SDK obtained after the login such the a user would need to login again. This at least keeps it consistent with the other services even though the access isn't removed strictly speaking.

Concerning your request about not showing the dialog when a user is already logged into his Microsoft account. The thing here is that we don't know before loading the authentication page if a user is already logged in which means that we have to present the dialog no matter what.

I will tackle the dialog text soon. But we can not put the name of the service into the message because the activity that handles authentication only knows the url that needs to be opened but not which service it is. So the only thing i could do is change it to sth like: "Connecting to service...".

So to sum this up: If the dialog text is changed this issue can be closed? Or is there anything else here?

**Comment by donpwilson** [comment]
Agreed. If the text is improved, the issue can be closed. But don't make it "service" users don't know about services. I use "Cloud" and "Cloud Storage" as generic terms.

**Comment by PatrickStoklasa** [comment]
The text should be better with the new version now. I found a rather easy way to include the service name into the text, as you mentioned. Let me know if this works for you.

**Comment by donpwilson** [comment]
This is now looking good!
